clients/drcachesim/simulator/caching_device_stats.h

```diff
@@ -64,9 +67,88 @@ enum class metric_name_t {
     FLUSHES
 };
 
+struct bound {
+    addr_t beg;
+    addr_t end;
+};
+
+class access_count_t {
+public:
+    access_count_t(int block_size)
+        : block_size_(block_size)
+    {
+        if (!IS_POWER_OF_2(block_size)) {
+            ERRMSG("Block size should be a power of 2.");
+            return;
+        }
+        int block_size_bits = compute_log2(block_size);
+        block_size_mask_ = ~((1 << block_size_bits) - 1);
+    }
+
+    // Takes non-alligned address and inserts bound consisting of the nearest multiples
+    // of the block_size.
+    void
+    insert(addr_t addr_beg, size_t index)
+    {
+        // Round the address down to the nearest multiple of the block_size.
+        addr_beg &= block_size_mask_;
+        addr_t addr_end = addr_beg + block_size_;
+
+        // Detect the overflow and assign maximum possible value to the addr_end.
+        if (addr_beg > addr_end) {
+            addr_end = std::numeric_limits<addr_t>::max();
+        }
+
+        if (index == bounds.size()) {
+            if (index > 0 && bounds[index - 1].end == addr_beg) {
+                bounds[index - 1].end = addr_end;
+            } else {
+                bounds.emplace_back(bound{ addr_beg, addr_end });
+            }
+            // Current bound -> (addr_beg...addr_end) connects bound[index] and
+            // bound[index - 1]
+        } else if (index > 0 && bounds[index - 1].end == addr_beg &&
+                   bounds[index].beg == addr_end) {
+            bounds[index - 1].end = bounds[index].end;
+            bounds.erase(bounds.begin() + index);
+            // Current bound extends bound[index - 1]
+        } else if (index > 0 && bounds[index - 1].end == addr_beg) {
+            bounds[index - 1].end = addr_end;
+            // Current bound extends bound[index]
+        } else if (bounds[index].beg == addr_end) {
+            bounds[index].beg = addr_beg;
+        } else {
+            bounds.insert(bounds.begin() + index, bound{ addr_beg, addr_end });
+        }
+    }
+
+    // Takes non-aligned address. Returns:
+    // - boolean value indicating whether the address has ever been accessed
+    // - index of the bound where the address is located/should be inserted.
+    std::pair<bool, size_t>
+    lookup(addr_t addr)
+    {
+        size_t i = 0;
+        for (const auto &bound : bounds) {
+            if (addr < bound.beg) {
+                return std::make_pair(false, i);
+            } else if (addr >= bound.beg && addr < bound.end) {
+                return std::make_pair(true, i);
+            }
+            i++;
+        }
+        return std::make_pair(false, bounds.size());
+    }
+
+private:
+    std::vector<bound> bounds;
```

Comment:
So this is an interval tree, right?  I would suggest storing it in a tree which should perform much better (as it is it looks like it will perform a whole lot of vector shifts).  You should be able to use std::map (esp std::map::upper_bound).

If this is changed to a tree does the overhead still show up enough for this to be off by default?

Reply:
I changed implementation to use std::map, it is way more effective indeed. The overhead is small enough that I turned this on by default. Thanks for the good suggestion!